internal/controller/webhook.go

```diff
@@ -0,0 +1,148 @@
+/*
+Copyright 2024 Red Hat OpenShift Data Foundation.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controller
+
+import (
+	admrv1 "k8s.io/api/admissionregistration/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	"k8s.io/utils/ptr"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+
+	odfwebhook "github.com/red-hat-storage/odf-operator/internal/webhook"
+)
+
+const (
+	WebhookServiceName = "odf-operator-webhook-service"
+	WebhookName        = "csv.odf.openshift.io"
+)
+
+func (r *StorageSystemReconciler) reconcileSubscriptionValidatingWebhook() error {
+
+	csvWebhook := admrv1.MutatingWebhook{
+		Name: WebhookName,
+		ClientConfig: admrv1.WebhookClientConfig{
+			Service: &admrv1.ServiceReference{
+				Name: WebhookServiceName,
+				Path: ptr.To(odfwebhook.WebhookPath),
+				Port: ptr.To(int32(443)),
+			},
+		},
+		Rules: []admrv1.RuleWithOperations{
+			{
+				Rule: admrv1.Rule{
+					APIGroups:   []string{"operators.coreos.com"},
+					APIVersions: []string{"v1alpha1"},
+					Resources:   []string{"clusterserviceversions"},
+					Scope:       ptr.To(admrv1.NamespacedScope),
+				},
+				Operations: []admrv1.OperationType{admrv1.Create},
+			},
+		},
+		SideEffects:             ptr.To(admrv1.SideEffectClassNone),
+		TimeoutSeconds:          ptr.To(int32(30)),
+		AdmissionReviewVersions: []string{"v1"},
+		// fail the validation if webhook can't be reached
+		FailurePolicy: ptr.To(admrv1.Fail),
+	}
```

Comment:
this whole part is basically a template which doesn't need to be instantiated everytime in a function and so pls move this to a templates package and mutate a deepcopy whenever required in createorupdate method.